From ffd81f212b5594b498f0ba07645c7a181540e494 Mon Sep 17 00:00:00 2001
From: Kiran Ramachandra <kiranmr@google.com>
Date: Fri, 21 Jun 2024 17:53:00 +0000
Subject: [PATCH] RESTRICT AUTOMERGE Fix to restrict admin from granting
 permission to a sensor permission group

Bug: 308138085
Test: atest PermissionControllerMockingTests:AdminRestrictedPermissionsUtilsTest

(cherry picked from commit 682c92ee1e47918993a860ddb1ce02277f6f5a8a)
Relnote: Security bug fix to restrict admin from granting permission to a sensor permission group

LOW_COVERAGE_REASON=b/330904893
(cherry picked from commit 7b959874bb348a2044bb7968e302f413a9f38d28)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:4857b928217abcbed4c24a180b6285a23174e565)
Merged-In: I7b0436565e2eb30eebd7f03375938ab6eba16951
Change-Id: I7b0436565e2eb30eebd7f03375938ab6eba16951
---
 .../PermissionControllerServiceImpl.java      |  3 +-
 .../v31/AdminRestrictedPermissionsUtils.java  | 10 +-
 PermissionController/tests/mocking/Android.bp |  1 +
 .../AdminRestrictedPermissionsUtilsTest.kt    | 91 +++++++++++++++++++
 4 files changed, 102 insertions(+), 3 deletions(-)
 create mode 100644 PermissionController/tests/mocking/src/com/android/permissioncontroller/tests/mocking/permission/utils/AdminRestrictedPermissionsUtilsTest.kt

diff --git a/PermissionController/src/com/android/permissioncontroller/permission/service/PermissionControllerServiceImpl.java b/PermissionController/src/com/android/permissioncontroller/permission/service/PermissionControllerServiceImpl.java
index d5c43335c..a7c52a531 100644
--- a/PermissionController/src/com/android/permissioncontroller/permission/service/PermissionControllerServiceImpl.java
+++ b/PermissionController/src/com/android/permissioncontroller/permission/service/PermissionControllerServiceImpl.java
@@ -563,7 +563,8 @@ public final class PermissionControllerServiceImpl extends PermissionControllerL
             switch (grantState) {
                 case PERMISSION_GRANT_STATE_GRANTED:
                     if (AdminRestrictedPermissionsUtils.mayAdminGrantPermission(perm.getName(),
-                            canAdminGrantSensorsPermissions, isManagedProfile, dpm)) {
+                            group.getName(), canAdminGrantSensorsPermissions, isManagedProfile,
+                            dpm)) {
                         perm.setPolicyFixed(true);
                         group.grantRuntimePermissions(false, false, new String[]{permName});
                         autoGrantPermissionsNotifier.onPermissionAutoGranted(permName);
diff --git a/PermissionController/src/com/android/permissioncontroller/permission/utils/v31/AdminRestrictedPermissionsUtils.java b/PermissionController/src/com/android/permissioncontroller/permission/utils/v31/AdminRestrictedPermissionsUtils.java
index 770ee6c95..4fde616e3 100644
--- a/PermissionController/src/com/android/permissioncontroller/permission/utils/v31/AdminRestrictedPermissionsUtils.java
+++ b/PermissionController/src/com/android/permissioncontroller/permission/utils/v31/AdminRestrictedPermissionsUtils.java
@@ -25,6 +25,7 @@ import android.os.UserManager;
 import android.util.ArraySet;
 
 import com.android.modules.utils.build.SdkLevel;
+import com.android.permissioncontroller.permission.utils.PermissionMapping;
 
 /**
  * A class for dealing with permissions that the admin may not grant in certain configurations.
@@ -78,7 +79,7 @@ public final class AdminRestrictedPermissionsUtils {
     /**
      * Returns true if the admin may grant this permission, false otherwise.
      */
-    public static boolean mayAdminGrantPermission(String permission,
+    public static boolean mayAdminGrantPermission(String permission, String permissionGroup,
             boolean canAdminGrantSensorsPermissions, boolean isManagedProfile,
             DevicePolicyManager dpm) {
         if (!SdkLevel.isAtLeastS()) {
@@ -87,7 +88,12 @@ public final class AdminRestrictedPermissionsUtils {
         if (isManagedProfile && Manifest.permission.READ_SMS.equals(permission)) {
             return mayManagedProfileAdminGrantReadSms(dpm);
         }
-        if (!ADMIN_RESTRICTED_SENSORS_PERMISSIONS.contains(permission)) {
+        boolean isAdminRestrictedSensorPermissionGroup = permissionGroup != null
+                && PermissionMapping.getPlatformPermissionNamesOfGroup(permissionGroup).stream()
+                .anyMatch(ADMIN_RESTRICTED_SENSORS_PERMISSIONS::contains);
+
+        if (!ADMIN_RESTRICTED_SENSORS_PERMISSIONS.contains(permission)
+                && !isAdminRestrictedSensorPermissionGroup) {
             return true;
         }
 
diff --git a/PermissionController/tests/mocking/Android.bp b/PermissionController/tests/mocking/Android.bp
index 64903d615..2d267eec5 100644
--- a/PermissionController/tests/mocking/Android.bp
+++ b/PermissionController/tests/mocking/Android.bp
@@ -113,6 +113,7 @@ android_test {
         "safety-label",
         "role-controller",
         "lottie",
+        "platform-test-annotations",
 
         "androidx.test.rules",
         "androidx.test.ext.truth",
diff --git a/PermissionController/tests/mocking/src/com/android/permissioncontroller/tests/mocking/permission/utils/AdminRestrictedPermissionsUtilsTest.kt b/PermissionController/tests/mocking/src/com/android/permissioncontroller/tests/mocking/permission/utils/AdminRestrictedPermissionsUtilsTest.kt
new file mode 100644
index 000000000..0c864da4a
--- /dev/null
+++ b/PermissionController/tests/mocking/src/com/android/permissioncontroller/tests/mocking/permission/utils/AdminRestrictedPermissionsUtilsTest.kt
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.permissioncontroller.tests.mocking.permission.utils
+
+import android.app.admin.DevicePolicyManager
+import android.platform.test.annotations.AsbSecurityTest
+import com.android.modules.utils.build.SdkLevel
+import com.android.permissioncontroller.permission.utils.v31.AdminRestrictedPermissionsUtils
+import org.junit.Assert.assertEquals
+import org.junit.Assume
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+import org.mockito.Mockito.mock
+
+@RunWith(Parameterized::class)
+class AdminRestrictedPermissionsUtilsTest(
+    private val permission: String,
+    private val group: String?,
+    private val canAdminGrantSensorsPermissions: Boolean,
+    private val expected: Boolean
+) {
+    private val dpm: DevicePolicyManager = mock(DevicePolicyManager::class.java)
+
+    @Before
+    fun setup() {
+        Assume.assumeTrue(SdkLevel.isAtLeastS())
+    }
+
+    @AsbSecurityTest(cveBugId = [308138085])
+    @Test
+    fun mayAdminGrantPermissionTest() {
+        val canGrant =
+            AdminRestrictedPermissionsUtils.mayAdminGrantPermission(
+                permission,
+                group,
+                canAdminGrantSensorsPermissions,
+                false,
+                dpm
+            )
+        assertEquals(expected, canGrant)
+    }
+
+    companion object {
+        /**
+         * Returns a list of arrays containing the following values:
+         *
+         * 0. Permission name (String)
+         * 1. Permission group name (String)
+         * 2. Can admin grant sensors permissions (Boolean)
+         * 3. Expected return from mayAdminGrantPermission method (Boolean)
+         */
+        @JvmStatic
+        @Parameterized.Parameters(name = "{index}: validate({0}, {1}, {3}) = {4}")
+        fun getParameters(): List<Array<out Any?>> {
+            return listOf(
+                arrayOf("abc", "xyz", false, true),
+                arrayOf("abc", null, false, true),
+                arrayOf("android.permission.RECORD_AUDIO", "xyz", false, false),
+                arrayOf("abc", "android.permission-group.MICROPHONE", false, false),
+                arrayOf(
+                    "android.permission.RECORD_AUDIO",
+                    "android.permission-group.MICROPHONE",
+                    false,
+                    false
+                ),
+                arrayOf(
+                    "android.permission.RECORD_AUDIO",
+                    "android.permission-group.MICROPHONE",
+                    true,
+                    true
+                ),
+            )
+        }
+    }
+}
-- 
2.46.1.824.gd892dcdcdd-goog

